src/Algebra/Graph/Relation/Symmetric.hs

```diff
@@ -11,27 +11,497 @@
+vertexList = R.vertexList . fromSymmetric
+
+-- | The sorted list of edges of a graph.
+-- Complexity: /O(n + m)/ time and /O(m)/ memory.
+--
+-- @
+-- edgeList 'empty'          == []
+-- edgeList ('vertex' x)     == []
+-- edgeList ('edge' x y)     == [(x,y), (y,x)]
+-- edgeList ('star' 2 [3,1]) == [(2,1), (1,2), (2,3), (3,2)]
+-- @
+edgeList :: SymmetricRelation a -> [(a, a)]
+edgeList = R.edgeList . fromSymmetric
+
+-- | The set of vertices of a given graph.
+-- Complexity: /O(1)/ time.
+--
+-- @
+-- vertexSet 'empty'      == Set.'Set.empty'
+-- vertexSet . 'vertex'   == Set.'Set.singleton'
+-- vertexSet . 'vertices' == Set.'Set.fromList'
+-- @
+vertexSet :: SymmetricRelation a -> Set.Set a
+vertexSet = R.vertexSet . fromSymmetric
+
+-- | The set of edges of a given graph.
+-- Complexity: /O(1)/ time.
+--
+-- @
+-- edgeSet 'empty'      == Set.'Set.empty'
+-- edgeSet ('vertex' x) == Set.'Set.empty'
+-- edgeSet ('edge' x y) == Set.'Set.fromList' [(x,y), (y,x)]
+-- @
+edgeSet :: SymmetricRelation a -> Set.Set (a, a)
+edgeSet = R.edgeSet . fromSymmetric
+
+-- | The sorted /adjacency list/ of a graph.
+-- Complexity: /O(n + m)/ time and /O(m)/ memory.
+--
+-- @
+-- adjacencyList 'empty'          == []
+-- adjacencyList ('vertex' x)     == [(x, [])]
+-- adjacencyList ('edge' 1 2)     == [(1, [2]), (2, [1])]
+-- adjacencyList ('star' 2 [3,1]) == [(1, [2]), (2, [1,3]), (3, [2])]
+-- 'stars' . adjacencyList        == id
+-- @
+adjacencyList :: Eq a => SymmetricRelation a -> [(a, [a])]
+adjacencyList = R.adjacencyList . fromSymmetric
+
+-- | The /preset/ of an element @x@ is the set of elements that are related to
+-- it on the /left/, i.e. @preSet x == { a | aRx }@. In the context of directed
+-- graphs, this corresponds to the set of /direct predecessors/ of vertex @x@.
+-- Complexity: /O(n + m)/ time and /O(n)/ memory.
+--
+-- @
+-- preSet x 'empty'      == Set.'Set.empty'
+-- preSet x ('vertex' x) == Set.'Set.empty'
+-- preSet 1 ('edge' 1 2) == Set.'Set.singleton' 2
+-- preSet y ('edge' x y) == Set.'Set.fromList' [x]
+-- @
+preSet :: Ord a => a -> SymmetricRelation a -> Set.Set a
+preSet x = R.preSet x . fromSymmetric
+
+-- | The /postset/ of an element @x@ is the set of elements that are related to
+-- it on the /right/, i.e. @postSet x == { a | xRa }@. In the context of directed
+-- graphs, this corresponds to the set of /direct successors/ of vertex @x@.
+-- Complexity: /O(n + m)/ time and /O(n)/ memory.
+--
+-- @
+-- postSet x 'empty'      == Set.'Set.empty'
+-- postSet x ('vertex' x) == Set.'Set.empty'
+-- postSet x ('edge' x y) == Set.'Set.fromList' [y]
+-- postSet 2 ('edge' 1 2) == Set.'Set.singleton' 1
+-- @
+postSet :: Ord a => a -> SymmetricRelation a -> Set.Set a
+postSet x = R.postSet x . fromSymmetric
+
+-- | The /path/ on a list of vertices.
+-- Complexity: /O((n + m) * log(n))/ time and /O(n + m)/ memory.
+--
+-- @
+-- path []        == 'empty'
+-- path [x]       == 'vertex' x
+-- path [x,y]     == 'edges' [(x,y), (y,x)]
```

Comment:
No need to duplicate the edge in this example.

Reply:
Already answered